File: 1155. Number of Dice Rolls With Target Sum.java
class Solution {
    public int numRollsToTarget(int n, int k, int target) {
        // Use a 2D array for dynamic programming
        int[][] dp = new int[n + 1][target + 1];
        int mod = 1000000007;

        // Base case: one way to achieve a target of 0 with 0 dice
        dp[0][0] = 1;

        // Fill in the dp array
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= target; j++) {
                // Calculate the current dp value based on the previous row
                for (int x = 1; x <= k && x <= j; x++) {
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - x]) % mod;
                }
            }
        }

        return dp[n][target];
    }
}

File: 1155. Number of Dice Rolls With Target Sum.cpp
class Solution {
public:
    int mod = 1e9 + 7;

    int solve(int n, int target, int k, std::vector<std::vector<int>> &memo) {
        if (target == 0 && n == 0) return 1;
        if (n == 0 || target < 0) return 0;

        if (memo[n][target] != -1) {
            return memo[n][target];
        }

        int ways = 0;
        for (int i = 1; i <= k; i++) {
            if (i <= target) {
                ways = (ways % mod + solve(n - 1, target - i, k, memo) % mod) % mod;
            }
        }

        memo[n][target] = ways;
        return ways;
    }

    int numRollsToTarget(int n, int k, int target) {
        // Initialize memoization table with -1
        std::vector<std::vector<int>> memo(n + 1, std::vector<int>(target + 1, -1));

        return solve(n, target, k, memo);
    }
};
